README.md
<h1 align="center" style="border-bottom: 2px solid #ebebeb;">CannonHttpJS</h1>
<div align="center" style="margin-bottom: 10px;">
<img src="https://img.shields.io/badge/license-MIT-green" />
<a href="https://codecov.io/gh/cannonmaster/CannonHttpJS" > 
 <img src="https://codecov.io/gh/cannonmaster/CannonHttpJS/branch/main/graph/badge.svg?token=ND97MPUCPP"/> 
 </a>
</div>

CannonHttpJS is a versatile HTTP client library that provides a simple and flexible interface for making HTTP requests in both Node.js and browser environments. It offers features such as request interceptors, response interceptors, caching, and timeout handling. With CannonHttpJS, you can easily handle RESTful APIs and communicate with servers for data retrieval and sending.

<h2 style="border-bottom: 2px solid #ebebeb;"> Key Features</h2>

- **HTTP Methods**: Perform common HTTP methods like GET, POST, PUT, PATCH, and DELETE.
- **Request Interceptors**: Intercept and modify request configurations before sending.
- **Response Interceptors**: Intercept and modify response data after receiving.
- **Caching**: Cache responses to improve performance and reduce duplicate requests.
- **Timeout Handling**: Set a timeout duration to ensure requests do not take too long.
- **Base URL**: Define a base URL to simplify the setting of request URLs.
- **Default Headers**: Set default headers to be applied to all requests.
- **Retry Mechanism**: Configure the number of retries and delay time for failed requests.

<h2 style="border-bottom: 2px solid #ebebeb;">Installation</h2>

```bash
npm install @cannonui/httpjs@latest
```

You're now ready to start using CannonHttpJS in your projects. Enjoy the power and flexibility it offers for handling HTTP requests!

<h2 style="border-bottom: 2px solid #ebebeb;">License</h2>

CannonHttpJS is open-source software licensed under the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
# Change Log

All notable changes to the CannonHttpJS library will be documented in this file.

## [1.0.0] - 2023-07-09

### Added

- Support for sending GET, POST, PUT, PATCH, and DELETE requests.
- Request interceptors to modify request configurations.
- Response interceptors to modify response data.
- Caching of responses for improved performance.
- Timeout handling to limit request duration.
- Base URL configuration for simplified URL setup.
- Default headers to apply to all requests.
- Retry mechanism to handle failed requests.
- Response data sanitization for data validation.

### Changed

- Renamed `request` method to `executeRequest`.
- Refactored internal request execution logic.
- Updated dependencies for improved compatibility.

### Fixed

- LRU cache strategy


test/client.test.ts
// @vitest-environment jsdom
import { s } from "vitest/dist/types-198fd1d9";
import CannonHttpJs, { RequestOptions, ResponseData } from "../src/main";
import { describe, it, expect, beforeEach, afterEach, vi } from "vitest";
import CannonHttp from "@cannonui/httpjs";
type RequestInterceptor<T> = (
  config: RequestOptions<T>
) => RequestOptions<T> | Promise<RequestOptions<T>>;

describe("CannonHttpJS", () => {
  let httpClient = CannonHttpJs;

  // production test
  // let httpClient = CannonHttp;

  beforeEach(() => {
    // httpClient = new CannonHttpJs();
    httpClient.clearRequestInterceptor();
    httpClient.clearResoponseInterceptor();
    httpClient.invalidateCache();
    httpClient.setCacheTime(0);
    httpClient.invalidateLocalStorage();
    vi.spyOn(httpClient, "executeRequest" as any);
  });

  it("should load json data", async () => {
    const localData = await httpClient.get("http://localhost:3000/json");
    const mockResponse: ResponseData<T> = {
      status: 200,
      statusText: "OK",
      headers: new Headers(),
      data: { a: 123 },
    };
    expect(JSON.stringify(localData)).toEqual(JSON.stringify(mockResponse));
  });

  it("should load data failed on first time and fetch success after save to the localstorage", async () => {
    const mockResponse: ResponseData<T> = {
      status: 200,
      statusText: "OK",
      headers: new Headers(),
      data: "hi",
    };
    // await httpClient.storeLocalData("http://localhost:3000/", mockResponse);

    const localData = await httpClient.get("http://localhost:3000/", {
      offline: true,
    });

    const data = await httpClient.getLocalData("http://localhost:3000/");

    expect(JSON.stringify(data)).toEqual(JSON.stringify(mockResponse));
  });

  it("should load data on key in localstorage", async () => {
    const mockResponse: ResponseData<T> = {
      status: 200,
      statusText: "OK",
      headers: new Headers(),
      data: "test data",
    };
    await httpClient.storeLocalData("http://localhost:3000/", mockResponse);

    const localData = await httpClient.get("http://localhost:3000/", {
      offline: true,
    });

    expect(JSON.stringify(localData)).toEqual(JSON.stringify(mockResponse));
  });
  it("should invalidate data on key in localstorage", async () => {
    httpClient.setBaseUrl("http://localhost:3000");
    const mockResponse: ResponseData<T> = {
      status: 200,
      statusText: "OK",
      headers: new Headers(),
      data: "test data",
    };
    await httpClient.storeLocalData("http://localhost:3000", mockResponse);

    httpClient.invalidateLocalStorage("http://localhost:3000");

    const localData = await httpClient.getLocalData("http://localhost:3000");

    expect(localData).toEqual(null);
  });

  it("should invalidate all data in localstorage", async () => {
    httpClient.setBaseUrl("http://localhost:3000");
    const mockResponse: ResponseData<T> = {
      status: 200,
      statusText: "OK",
      headers: new Headers(),
      data: "test data",
    };
    await httpClient.storeLocalData("http://localhost:3000", mockResponse);

    httpClient.invalidateLocalStorage();

    const localData = await httpClient.getLocalData("http://localhost:3000");

    expect(localData).toEqual(null);
  });

  it("should store data in localstorage", async () => {
    httpClient.setBaseUrl("http://localhost:3000");
    const mockResponse: ResponseData<T> = {
      status: 200,
      statusText: "OK",
      headers: new Headers(),
      data: "test data",
    };
    await httpClient.storeLocalData("http://localhost:3000", mockResponse);

    const localData = await httpClient.getLocalData("http://localhost:3000");

    expect(JSON.stringify(localData)).toEqual(JSON.stringify(mockResponse));
  });

  it("should get formdata if the isFormData is set to true", async () => {
    const data = { a: 123, b: "012" };

    httpClient.setBaseUrl("http://localhost:3000");

    const response = (await httpClient.post("/is-form-data", {
      data,
      isFormData: true,
    })) as ResponseData<string>;
    // Assertions
    expect(response.data.includes("multipart/form-data;")).toBe(true);
  });

  it("should get oldest cache entry", async () => {
    httpClient.setBaseUrl("http://localhost:3000");
    httpClient.setCacheSize(10);
    httpClient.setCacheTime(100000);
    await httpClient.get("/");
    await httpClient.get("/default-header");
    await httpClient.get("/endpoint");

    const cache = httpClient.getOldestEntry();

    // Assertions
    expect(cache).toBe("http://localhost:3000/");
  });

  it("should set multipart content type if data is formData", async () => {
    const data = new FormData();
    data.append("a", "123");
    httpClient.setBaseUrl("http://localhost:3000");
    const response = (await httpClient.post("/form-data", {
      data,
    })) as ResponseData<string>;
    // Assertions
    expect(response.data.includes("multipart/form-data;")).toBe(true);
  });

  it("should throw an error when request interceptor fails", async () => {
    const requestInterceptor: RequestInterceptor<any> = async () => {
      throw new Error("Request interceptor failed");
    };

    httpClient.addRequestInterceptor(requestInterceptor);

    const config: RequestOptions<any> = {
      url: "https://example.com",
    };

    try {
      await httpClient["executeRequest"](config);
      // Fail the test if no error is thrown
      expect(true).toBe(false);
    } catch (error) {
      expect(error.message).toContain("Request interceptor failed");
    }
  });

  it("should delay execution for the specified time", async () => {
    const delayTime = 1000; // 1 second

    const startTime = Date.now();
    await httpClient["delay"](delayTime);
    const endTime = Date.now();

    const elapsedTime = endTime - startTime;
    expect(elapsedTime).toBeGreaterThanOrEqual(delayTime);
  });

  it("should calculate the retry delay correctly", () => {
    const retryCount = 3;
    const expectedDelay = 4000; // 500 * 2^3 = 4000

    const delay = httpClient["calculateRetryDelay"](retryCount);
    expect(delay).toEqual(expectedDelay);
  });

  it("should send a GET request successfully", async () => {
    // Set the base URL
    httpClient.setBaseUrl("http://localhost:3000");

    // Make the GET request
    const response = await httpClient.get("/get-endpoint", {
      params: { key1: "value1", key2: "value2" },
    });

    // Assertions
    expect(response.status).toBe(200);
    expect(response.data).toBe("GET response");
  });

  it("should throw error on invalid data formmat", async () => {
    // Make the POST request
    const data = "123";

    // Assertions
    await expect(
      httpClient.post("http://localhost:3000/endpoint", {
        data,
      })
    ).rejects.toThrowError("invalid data format");
  });

  it("should send a POST request successfully", async () => {
    // Make the POST request
    const data = { a: 123 };
    const response = await httpClient.post("http://localhost:3000/endpoint", {
      data,
    });

    // Assertions
    expect(response.status).toBe(200);
    expect(response.data).toBe("post response");
  });

  it("should handle request interceptors", async () => {
    // Set the base URL
    httpClient.setBaseUrl("http://localhost:3000");

    // Add a request interceptor
    httpClient.addRequestInterceptor((config) => {
      // Modify the request config
      config.headers = {
        ...config.headers,
        Authorization: "Bearer token",
      };
      return config;
    });

    // Make the GET request
    const response = await httpClient.get("/auth-token");

    // Assertions
    expect(response.status).toBe(200);
    expect(response.data).toBe("auth token detected");
  });

  it("should invalidate the cache and send a new GET request successfully", async () => {
    // Set the base URL
    httpClient.setBaseUrl("http://localhost:3000");

    // Make the initial GET request
    const initialResponse = await httpClient.get("/cache");

    // Assertions for the initial request
    expect(initialResponse.status).toBe(200);
    expect(initialResponse.data).toBe("Init connection");
    // Invalidate the cache
    httpClient.invalidateCache("/cache");

    // Make the GET request again
    const responseAfterInvalidation = await httpClient.get("/cache");

    // Assertions for the subsequent request
    expect(responseAfterInvalidation.status).toBe(200);
    expect(responseAfterInvalidation.data).toBe("Init connection");
  });

  it("should return data from the cache ", async () => {
    // Set the base URL
    httpClient.setBaseUrl("http://localhost:3000");
    httpClient.setCacheSize(10);
    httpClient.setCacheTime(10000000);

    // Make the initial GET request
    const initialResponse = await httpClient.get("/cache");

    // Assertions for the initial request
    expect(initialResponse.status).toBe(200);
    expect(initialResponse.data).toBe("Init connection");
    initialResponse.data = "from cache";

    // // // Make the GET request again
    const responseAfterInvalidation = await httpClient.get("/cache");

    // // // Assertions for the subsequent request
    expect(responseAfterInvalidation.status).toBe(200);
    expect(responseAfterInvalidation.data).toBe("from cache");
  });

  it("should invalidate all cache", async () => {
    // Set the base URL
    httpClient.setBaseUrl("http://localhost:3000");

    // Make the GET request
    const response = await httpClient.get("/endpoint");

    // Assertions
    expect(response.status).toBe(200);
    expect(response.data).toBe("get response");

    // Invalidate all cache
    httpClient.invalidateCache();

    // Make the GET request again
    const responseAfterInvalidation = await httpClient.get("/endpoint");

    // Ensure the response is not retrieved from the cache
    expect(responseAfterInvalidation.status).toBe(200);
    expect(responseAfterInvalidation.data).toBe("get response");
  });

  it("should handle PATCH request", async () => {
    // Make the PATCH request
    const response = await httpClient.patch("http://localhost:3000/endpoint", {
      key: "value",
    });

    // Assertions
    expect(response.status).toBe(200);
    expect(response.data).toBe("patch response");
  });
  it("should handle request timeout", async () => {
    // Set the base URL
    httpClient.setBaseUrl("http://localhost:3000");

    // Make the GET request with a timeout of 500 milliseconds
    const timeout = 500;
    const requestOptions = { url: "/delay", timeout }; // Include the `url` and `timeout` properties in RequestOptions

    // Ensure the request throws an error due to timeout
    await expect(() =>
      httpClient.get("/delay", requestOptions)
    ).rejects.toThrowError(/Request timed out after \d+ms/);
  });

  it("should handle request timeout with success", async () => {
    // Set the base URL
    httpClient.setBaseUrl("http://localhost:3000");

    // Make the GET request with a timeout of 500 milliseconds
    const timeout = 3000;
    const requestOptions = { url: "/delay", timeout }; // Include the `url` and `timeout` properties in RequestOptions

    // Ensure the request throws an error due to timeout
    const res = await httpClient.get("/delay", requestOptions);
    expect(res.status).toBe(200);
    expect(res.data).toBe("timeout working");
  });

  it("should append query parameters to the URL", async () => {
    // Set the base URL
    httpClient.setBaseUrl("http://localhost:3000");

    // Make the GET request with query parameters
    const response = await httpClient.get("/param", {
      params: { key1: "value1", key2: "value2" },
    });

    // Assertions
    expect(response.status).toBe(200);
    expect(response.data).toBe("/param?key1=value1&key2=value2");
  });

  it("should handle PUT request", async () => {
    // Make the PUT request
    const response = await httpClient.put("http://localhost:3000/endpoint", {
      key: "value",
    });

    // Assertions
    expect(response.status).toBe(200);
    expect(response.data).toBe("put response");
  });

  it("should handle DELETE request", async () => {
    // Make the DELETE request
    const response = await httpClient.delete("http://localhost:3000/endpoint");

    // Assertions
    expect(response.status).toBe(200);
    expect(response.data).toBe("delete response");
  });

  it("should handle response interceptors", async () => {
    // Add a response interceptor
    httpClient.addResponseInterceptor((response) => {
      // Modify the response data
      response.data = "Modified response";
      return response;
    });

    // Make the GET request
    const response = await httpClient.get("http://localhost:3000/endpoint");

    // Assertions
    expect(response.status).toBe(200);
    expect(response.data).toBe("Modified response");
  });

  it("should handle default headers", async () => {
    // Set default headers
    httpClient.setDefaultHeaders({
      "Content-Type": "application/json",
      "x-custom": "x",
    });

    // Make the GET request
    const response = await httpClient.get(
      "http://localhost:3000/default-header"
    );

    // Assertions
    expect(response.status).toBe(200);
    expect(response.data).toBe("x");
  });

  it("should handle response sanitizers", async () => {
    // Add a response sanitizer
    httpClient.addResponseSanitizer((data: any) => {
      // Sanitize the response data
      return "Sanitized data";
    });

    // Make the GET request
    const response = await httpClient.get("http://localhost:3000/endpoint");

    // Assertions
    expect(response.status).toBe(200);
    expect(response.data).toBe("Sanitized data");
  });

  it("should set cache size", () => {
    const cacheSize = 10;
    httpClient.setCacheSize(cacheSize);

    // Get the cache size
    const currentCacheSize = httpClient["cacheSize"];

    // Assertion
    expect(currentCacheSize).toBe(cacheSize);
  });

  it("should set retry count", () => {
    const maxRetry = 3;
    httpClient.setRetry(maxRetry);

    // Get the retry count
    const currentMaxRetry = httpClient["maxRetry"];

    // Assertion
    expect(currentMaxRetry).toBe(maxRetry);
  });

  it("should set cache time", () => {
    const cacheTime = 60000;
    httpClient.setCacheTime(cacheTime);

    // Get the default cache time
    const currentCacheTime = httpClient["defaultCacheTime"];

    // Assertion
    expect(currentCacheTime).toBe(cacheTime);
  });

  it("should retry the request based on the maxRetry configuration", async () => {
    // Set the maxRetry value to 3
    const maxRetry = 3;
    httpClient.setRetry(maxRetry);

    // Make the GET request
    try {
      await httpClient.get("http://localhost:3000/max");
    } catch (error) {
      // Ensure the error message matches the expected format

      expect(error.message).toMatch(
        /^Request failed.*and the request was aborted$/s
      );

      // Ensure the 'executeRequest' method was called the expected number of times
      expect(httpClient["executeRequest"]).toHaveBeenCalledTimes(maxRetry + 1);
    }
  });

  it("should get the right cache key after renmove the oldest cache key", async () => {
    // Set the max cache value to 2
    httpClient.setCacheSize(2);
    httpClient.setCacheTime(100000);

    await httpClient.get("http://localhost:3000/cache");
    await httpClient.get("http://localhost:3000/");
    await httpClient.get("http://localhost:3000/auth-token");

    expect(httpClient.getOldestEntry()).toBe("http://localhost:3000/");
    expect(httpClient.cache.size).toBe(2);
  });

  // it("should formdata handle file success", async () => {
  //   const formData = {
  //     a: 123,
  //     file: new File(["Hello, World!"], "myfile.txt", { type: "text/plain" }),
  //   };
  //   const formData = new FormData();
  //   formData.append("a", "123");
  //   formData.append(
  //     "file",
  //     new File(["Hello, World!"], "myfile.txt", { type: "text/plain" })
  //   );
  //   fetch("http://localhost:3000/file", {
  //     method: "POST",
  //     body: formData,
  //   }).catch((err) => {
  //     console.log(err);
  //   });
  //   httpClient.setBaseUrl("http://localhost:3000");
  //   const response = await httpClient.post("/file", {
  //     data: formData,
  //     isFormData: true,
  //   });
  //   // Assertions
  //   expect(response.data).toBe("abc");
  // });

  afterEach(() => {});
});


test/server.test.ts
import CannonHttpJs, {
  RequestOptions,
  ResponseData,
  ExtendedRequestOptions,
} from "../src/main";
import { describe, it, expect, beforeEach, afterEach, vi } from "vitest";
import nock, { RequestBodyMatcher } from "nock";
import CannonHttp from "@cannonui/httpjs";
// const CannonHttp = require("@cannonui/httpjs");
type RequestInterceptor<T> = (
  config: RequestOptions<T>
) => RequestOptions<T> | Promise<RequestOptions<T>>;

describe("CannonHttpJS", () => {
  // let httpClient = CannonHttpJs;

  // production
  let httpClient = CannonHttp;

  beforeEach(() => {
    // httpClient = new CannonHttpJs();
    httpClient.clearRequestInterceptor();
    httpClient.clearResoponseInterceptor();
    httpClient.invalidateCache();
    httpClient.setCacheTime(0);
    vi.spyOn(httpClient, "executeRequest" as any);
  });

  it("should throw an error when request interceptor fails", async () => {
    const requestInterceptor: RequestInterceptor<any> = async () => {
      throw new Error("Request interceptor failed");
    };

    httpClient.addRequestInterceptor(requestInterceptor);

    const config: ExtendedRequestOptions<any> = {
      url: "https://example.com",
    };

    try {
      await httpClient["executeRequest"](config);
      // Fail the test if no error is thrown
      expect(true).toBe(false);
    } catch (error) {
      expect(error.message).toContain("Request interceptor failed");
    }
  });

  it("should delay execution for the specified time", async () => {
    const delayTime = 1000; // 1 second

    const startTime = Date.now();
    await httpClient["delay"](delayTime);
    const endTime = Date.now();

    const elapsedTime = endTime - startTime;
    expect(elapsedTime).toBeGreaterThanOrEqual(delayTime);
  });

  it("should calculate the retry delay correctly", () => {
    const retryCount = 3;
    const expectedDelay = 4000; // 500 * 2^3 = 4000

    const delay = httpClient["calculateRetryDelay"](retryCount);
    expect(delay).toEqual(expectedDelay);
  });

  it("should send a GET request successfully", async () => {
    // Set the base URL
    httpClient.setBaseUrl("https://api.example.com");

    // Mock the URL
    // const scope = nock("https://api.example.com")
    //   .get("/endpoint")
    //   .reply(200, "GET response");

    // Make the GET request
    const response = await httpClient.get("http://localhost:3000/get-endpoint");

    // Assertions
    expect(response.status).toBe(200);
    expect(response.data).toBe("GET response");

    // Ensure the mock URL was called
    // expect(scope.isDone()).toBe(true);
  });

  it("should send a POST request successfully", async () => {
    // Mock the URL
    // const scope = nock("https://api.example.com")
    //   .post("/endpoint")
    //   .reply(201, "POST response");

    // Make the POST request
    const response = await httpClient.post("http://localhost:3000/endpoint", {
      data: { a: 123 },
    });

    // Assertions
    expect(response.status).toBe(200);
    expect(response.data).toBe("post response");

    // Ensure the mock URL was called
    // expect(scope.isDone()).toBe(true);
  });

  it("should handle request interceptors", async () => {
    // Set the base URL
    httpClient.setBaseUrl("http://localhost:3000");

    // Mock the URL
    // const scope = nock("https://api.example.com")
    //   .get("/endpoint")
    //   .reply(200, "GET response");

    // Add a request interceptor
    httpClient.addRequestInterceptor((config: ExtendedRequestOptions<any>) => {
      // Modify the request config
      config.headers = {
        ...config.headers,
        Authorization: "Bearer token",
      } as Record<string, string>;
      return config;
    });

    // Make the GET request
    const response = await httpClient.get("/endpoint");

    // Assertions
    expect(response.status).toBe(200);
    expect(response.data).toBe("get response");

    // Ensure the mock URL was called
    // expect(scope.isDone()).toBe(true);
    // Ensure the request interceptor was applied
    // expect(scope.pendingMocks().length).toBe(0);
  });

  it("should invalidate the cache and send a new GET request successfully", async () => {
    // Set the base URL
    httpClient.setBaseUrl("http://localhost:3000");

    // Mock the URL for the initial request
    // const scope = nock("https://api.example.com")
    //   .get("/endpoint")
    //   .reply(200, "GET response");

    // Make the initial GET request
    const initialResponse = await httpClient.get("/cache");

    // Assertions for the initial request
    expect(initialResponse.status).toBe(200);
    expect(initialResponse.data).toBe("Init connection");
    // expect(scope.isDone()).toBe(true);

    // Invalidate the cache
    httpClient.invalidateCache("/cache");
    const testHeaders = new Headers();
    testHeaders.append("Conttent-Type", "application/json;");
    httpClient["cache"].set("http://localhost:3000/cache", {
      data: {
        status: 200,
        statusText: "abc",
        headers: testHeaders,
        data: "cached connection",
      },
      expiresAt: Date.now(),
    });

    // Mock the URL for the subsequent request
    // const scopeAfterInvalidation = nock("https://api.example.com")
    //   .get("/endpoint")
    //   .reply(200, "New GET response");

    // Make the GET request again
    const responseAfterInvalidation = await httpClient.get("/cache");

    // Assertions for the subsequent request
    expect(responseAfterInvalidation.status).toBe(200);
    expect(responseAfterInvalidation.data).toBe("Init connection");
    // expect(scopeAfterInvalidation.isDone()).toBe(true);
  });
  it("should invalidate all cache", async () => {
    // Set the base URL
    httpClient.setBaseUrl("http://localhost:3000");

    // Mock the URL
    // const scope = nock("https://api.example.com")
    //   .get("/endpoint")
    //   .reply(200, "GET response");

    // Make the GET request
    const response = await httpClient.get("/endpoint");

    // Assertions
    expect(response.status).toBe(200);
    expect(response.data).toBe("get response");

    // Ensure the mock URL was called
    // expect(scope.isDone()).toBe(true);

    // Invalidate all cache
    httpClient.invalidateCache();

    // Mock the URL for the subsequent request
    // const scopeAfterInvalidation = nock("https://api.example.com")
    //   .get("/endpoint")
    //   .reply(200, "GET response");

    // Make the GET request again
    const responseAfterInvalidation = await httpClient.get("/endpoint");

    // Ensure the response is not retrieved from the cache
    expect(responseAfterInvalidation.status).toBe(200);
    expect(responseAfterInvalidation.data).toBe("get response");

    // Ensure the mock URL is called again
    // expect(scope.pendingMocks().length).toBe(0);
  });

  it("should handle PATCH request", async () => {
    // Mock the URL with key-value matching
    // const scope = nock("https://api.example.com")
    //   .patch("/endpoint", { key: "value" })
    //   .reply(200, "PATCH response");

    // Make the PATCH request
    const response = await httpClient.patch("http://localhost:3000/endpoint", {
      data: { key: "value" },
    });

    // Assertions
    expect(response.status).toBe(200);
    expect(response.data).toBe("patch response");

    // Ensure the mock URL was called
    // expect(scope.isDone()).toBe(true);
  });
  it("should handle request timeout", async () => {
    // Set the base URL
    httpClient.setBaseUrl("http://localhost:3000");

    // Mock the URL
    // const scope = nock("https://api.example.com")
    //   .get("/endpoint")
    //   .delay(1000) // Delay the response by 1 second
    //   .reply(200, "GET response");

    // Make the GET request with a timeout of 500 milliseconds
    const timeout = 500;
    const requestOptions = { url: "/endpoint", timeout }; // Include the `url` and `timeout` properties in RequestOptions

    // Ensure the request throws an error due to timeout
    await expect(() =>
      httpClient.get("/delay", requestOptions)
    ).rejects.toThrowError(/Request timed out after \d+ms/);

    // Ensure the mock URL was called
    // expect(scope.isDone()).toBe(true);
  });

  it("should append query parameters to the URL", async () => {
    // Set the base URL
    httpClient.setBaseUrl("http://localhost:3000");

    // Mock the URL
    // const scope = nock("https://api.example.com")
    //   .get("/endpoint?key1=value1&key2=value2")
    //   .reply(200, "GET response");

    // Make the GET request with query parameters
    const response = await httpClient.get("/endpoint", {
      params: { key1: "value1", key2: "value2" },
    });

    // Assertions
    expect(response.status).toBe(200);
    expect(response.data).toBe("get response");

    // Ensure the mock URL was called
    // expect(scope.isDone()).toBe(true);
  });

  it("should handle PUT request", async () => {
    // Mock the URL with key-value matching
    // const scope = nock("https://api.example.com")
    //   .put("/endpoint", { key: "value" })
    //   .reply(200, "PUT response");

    // Make the PUT request
    const response = await httpClient.put("http://localhost:3000/endpoint", {
      data: { key: "value" },
    });

    // Assertions
    expect(response.status).toBe(200);
    expect(response.data).toBe("put response");

    // Ensure the mock URL was called
    // expect(scope.isDone()).toBe(true);
  });

  it("should handle DELETE request", async () => {
    // Mock the URL
    // const scope = nock("https://api.example.com")
    //   .delete("/endpoint")
    //   .reply(200, "DELETE response");

    // Make the DELETE request
    const response = await httpClient.delete("http://localhost:3000/endpoint");

    // Assertions
    expect(response.status).toBe(200);
    expect(response.data).toBe("delete response");

    // Ensure the mock URL was called
    // expect(scope.isDone()).toBe(true);
  });

  it("should handle response interceptors", async () => {
    // Mock the URL
    // const scope = nock("https://api.example.com")
    //   .get("/endpoint")
    //   .reply(200, "GET response");

    // Add a response interceptor
    httpClient.addResponseInterceptor((response) => {
      // Modify the response data
      response.data = "Modified response";
      return response;
    });

    // Make the GET request
    const response = await httpClient.get("http://localhost:3000/endpoint");

    // Assertions
    expect(response.status).toBe(200);
    expect(response.data).toBe("Modified response");

    // Ensure the mock URL was called
    // expect(scope.isDone()).toBe(true);
  });

  it("should handle response sanitizers", async () => {
    // Mock the URL
    // const scope = nock("https://api.example.com")
    //   .get("/endpoint")
    //   .reply(200, { data: "Original data" });

    // Add a response sanitizer
    httpClient.addResponseSanitizer((data: any) => {
      // Sanitize the response data
      return "Sanitized data";
    });

    // Make the GET request
    const response = await httpClient.get("http://localhost:3000/endpoint");

    // Assertions
    expect(response.status).toBe(200);
    expect(response.data).toBe("Sanitized data");

    // Ensure the mock URL was called
    // expect(scope.isDone()).toBe(true);
  });

  it("should set cache size", () => {
    const cacheSize = 10;
    httpClient.setCacheSize(cacheSize);

    // Get the cache size
    const currentCacheSize = httpClient["cacheSize"];

    // Assertion
    expect(currentCacheSize).toBe(cacheSize);
  });

  it("should set retry count", () => {
    const maxRetry = 3;
    httpClient.setRetry(maxRetry);

    // Get the retry count
    const currentMaxRetry = httpClient["maxRetry"];

    // Assertion
    expect(currentMaxRetry).toBe(maxRetry);
  });

  it("should set cache time", () => {
    const cacheTime = 60000;
    httpClient.setCacheTime(cacheTime);

    // Get the default cache time
    const currentCacheTime = httpClient["defaultCacheTime"];

    // Assertion
    expect(currentCacheTime).toBe(cacheTime);
  });

  it("should retry the request based on the maxRetry configuration", async () => {
    // Set the maxRetry value to 3
    const maxRetry = 3;
    httpClient.setRetry(maxRetry);

    // Set up a mock URL that always returns an error
    // const scope = nock("https://api.example.com")
    //   .get("/endpoint")
    //   .replyWithError("Request error");

    // Make the GET request
    try {
      await httpClient.get("http://localhost:3000/endpoint");
    } catch (error) {
      // Ensure the error message matches the expected format

      expect(error.message).toMatch(
        /^Request failed.*and the request was aborted$/s
      );

      // Ensure the 'executeRequest' method was called the expected number of times
      expect(httpClient["executeRequest"]).toHaveBeenCalledTimes(maxRetry + 1);

      // Ensure the mock URL was called the expected number of times
      // expect(scope.isDone()).toBe(true);
    }
  });

  afterEach(() => {
    nock.cleanAll();
  });
});


backend/app.js
import express from "express";
import cors from "cors";
import { StatusCodes } from "http-status-codes";
import multer from "multer";
// import httpjs from "@cannonui/httpjs";
import morgan from "morgan";
const storage = multer.memoryStorage();
const upload = multer({ dest: "uploads/" });
const app = express();
let cache = false;
app.use(express.json());
app.use(
  express.urlencoded({
    extended: true,
  })
);
// app.get("/server", async (req, res) => {
//   const ok = await httpjs.get("http://localhost:3000");
//   console.log(ok);

//   res.status(StatusCodes.OK).send("ok");
// });
// app.use(morgan("dev"));
app.use(cors());
app.get("/auth-token", (req, res) => {
  res.status(StatusCodes.OK).send("auth token detected");
});
app.get("/get-endpoint", (req, res) => {
  res.status(StatusCodes.OK).send("GET response");
});
app.get("/", (req, res) => {
  // console.log("abc");
  res.status(StatusCodes.OK).send("hi");
});
app.get("/default-header", (req, res) => {
  // console.log(req.headers);
  res.status(StatusCodes.OK).send(req.headers["x-custom"]);
});
app.get("/endpoint", (req, res) => {
  res.status(StatusCodes.OK).send("get response");
});
app.get("/param", (req, res) => {
  res.status(StatusCodes.OK).send(req.url);
});
app.get("/cache", (req, res) => {
  res.status(StatusCodes.OK).send("Init connection");
});
app.get("/json", (req, res) => {
  res.status(StatusCodes.OK).json({ a: 123 });
});
app.get("/max", (req, res) => {
  res.status(StatusCodes.BAD_REQUEST).send("testing");
});
app.get("/delay", async (req, res) => {
  const delay = async (seconds) => {
    return new Promise((resolve, reject) => {
      setTimeout(() => {
        res.status(StatusCodes.OK).send("timeout working");
        resolve();
      }, seconds);
    });
  };
  await delay(2000);
});

app.patch("/endpoint", (req, res) => {
  res.status(StatusCodes.OK).send("patch response");
});
app.put("/endpoint", (req, res) => {
  res.status(StatusCodes.OK).send("put response");
});
app.delete("/endpoint", (req, res) => {
  res.status(StatusCodes.OK).send("delete response");
});

app.post("/endpoint", (req, res) => {
  res.status(StatusCodes.OK).send("post response");
});

app.post("/form-data", (req, res) => {
  // console.log(req.body);
  res.status(StatusCodes.OK).send(req.headers["content-type"]);
});

app.post("/is-form-data", upload.none(), (req, res) => {
  res.status(StatusCodes.OK).send(req.headers["content-type"]);
});

app.post("/file", upload.single("file"), (req, res) => {
  // console.log(req.headers, "---------");
  // console.log(req.body);
  // console.log(req.data);
  // console.log(req.file, "abc");
  res.status(StatusCodes.OK).send("abc");
});

app.listen(3000, () => {
  console.log("server running on port 3000");
});
